### PR TITLE
mmap file

### PR DESCRIPTION
**Pull request by cloudbank**
I added a memory mapped file option for the model file.  Since the model files have to be copied out of the apk due to assets directory readonly status and inability to create the FileChannel directly, it is assumed that you have done so to internal storage with the same context that is passed into the overloaded constructor.  Internal storage eliminates problems of SD card not being present or M runtime permissions requests but is also in danger of being full etc.  Expansion files are the only other option, but this requires a download from the play store, which is an extra step.  I therefore did the most simple thing.

### CODE REVIEW

**Comment by cloudbank** [comment]
I could not squash my commits so I created a new request with a clean history.

**Comment by yifeif** [comment]
@andrewharp could you take a look or reassign? Thank you!

**Comment by cloudbank** [comment]
@andrewharp if you want some code that uses it and copies files initially,  I am working on a library  which is using the TensorFlowImageClassifier. I see the comment:
TODO(andrewharp): make this handle non-assets.
If you have ideas I have not mentioned about where to hold the model files, I would be flexible about working on some other solutions too.

File: tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java [hunk 1]
```diff
@@ -15,40 +15,114 @@
 
 package org.tensorflow.contrib.android;
 
+import android.content.Context;
```
Comment (andrewharp) [inline tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: Please revert the import reordering

File: tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java [hunk 2]
```diff
@@ -118,6 +192,7 @@ public TensorFlowInferenceInterface(InputStream is) {
    *
    * @param outputNames A list of output nodes which should be filled by the inference pass.
    */
+
```
Comment (andrewharp) [inline tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: Can you make a pass and delete errant blank lines like this please? No blank lines are necessary after comments or after opening braces or before closing braces, and a single blank line is preferred after closing braces (depending on context).

File: tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java [hunk 3]
```diff
@@ -458,6 +537,39 @@ private void prepareNativeRuntime() {
...
+    try {
+      g.importGraphDef(graphDef);
+    } catch (IllegalArgumentException e) {
+      throw new IOException("Not a valid TensorFlow Graph serialization: " + e.getMessage());
+    }
+
+    if (VERSION.SDK_INT >= 18) {
+      Trace.endSection(); // importGraphDef.
+      Trace.endSection(); // initializeTensorFlow.
```
Comment (andrewharp) [inline tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: This seems to match up with "readGraphDef", not "initializeTensorFlow".

In general the tracing is a good idea, but it should probably be added consistently to the entire interface in a separate PR.

**Comment by andrewharp** [comment]
@cloudbank That TODO refers to the label file for the demo specifically. Since that's just example app code (under examples/android) rather than library code (under contrib/android) it felt less urgent to generalize it.

I would not be opposed to having the TFII class do a check and copy the file from assets to a specified tmp dir if not already present, if we can show that this sufficiently speeds up subsequent runs.

**Comment by cloudbank** [comment]
@andrewharp  The worry point is when to delete the file that was copied. Although this does not affect the buffer, obviously the mapped file won't be there forever either. Ideally, this heavy create operation involving the big graph file is done off heap avoiding IO exactly once and the classifier object persisted for further usage. Of course all of this is taken care of outside of the TFII in my work.   I certainly could add copy and delete methods if you insist but would not want to encourage repeated copy, or for that matter repeated create.  I would love to work on adding persistence and feel safe to copy/delete once and then null out the buffer.  
I am fixing the typo and whitespace edits.  Tell me what you think about the delete.

**Pull request closed without merging**